test/token_holder/execute_rule.js

```diff
@@ -0,0 +1,605 @@
+                mockRuleValue,
+            );
+            const nonce = 1;
+
+            const [, rsv] = getExecuteRuleExTxData(
+                tokenHolder.address,
+                mockRule.address,
+                mockRulePassActionData,
+                nonce,
+                ephemeralPrivateKey1,
+            );
+
+            await utils.expectRevert(
+                tokenHolder.executeRule(
+                    mockRule.address,
+                    mockRulePassActionData,
+                    nonce,
+                    rsv.v,
+                    ethUtils.bufferToHex(rsv.r),
+                    ethUtils.bufferToHex(rsv.s),
+                ),
+                'Transaction is signed with non authorized key.',
+            );
+        });
+
+        it('Invalid nonce.', async () => {
+            const spendingLimit = 10;
+            const deltaExpirationHeight = 50;
+            const tokenHolder = await createTokenHolder(
+                accounts[0],
+                ephemeralKeyAddress1,
+                spendingLimit,
+                deltaExpirationHeight,
+            );
+
+            const mockRule = await MockRule.new();
+            const mockRuleValue = accounts[1];
+            const mockRuleSucceedActionData = generateMockRulePassActionData(
+                mockRuleValue,
+            );
+
+            // Correct nonce is 1.
+            const invalidNonce0 = 0;
+
+            const [, rsv0] = getExecuteRuleExTxData(
+                tokenHolder.address,
+                mockRule.address,
+                mockRuleSucceedActionData,
+                invalidNonce0, // correct nonce is 1
+                ephemeralPrivateKey2,
+            );
+
+            await utils.expectRevert(
+                tokenHolder.executeRule(
+                    mockRule.address,
+                    mockRuleSucceedActionData,
+                    invalidNonce0,
+                    rsv0.v,
+                    ethUtils.bufferToHex(rsv0.r),
+                    ethUtils.bufferToHex(rsv0.s),
+                ),
+                'Transaction is signed with non authorized key.',
+            );
+
+            // correct nonce is 1.
+            const invalidNonce2 = 2;
+
+            const [, rsv2] = getExecuteRuleExTxData(
+                tokenHolder.address,
+                mockRule.address,
+                mockRuleSucceedActionData,
+                invalidNonce2, // correct nonce is 1
+                ephemeralPrivateKey2,
+            );
+
+            await utils.expectRevert(
+                tokenHolder.executeRule(
+                    mockRule.address,
+                    mockRuleSucceedActionData,
+                    invalidNonce2,
+                    rsv2.v,
+                    ethUtils.bufferToHex(rsv2.r),
+                    ethUtils.bufferToHex(rsv2.s),
+                ),
+                'Transaction is signed with non authorized key.',
+            );
+        });
+    });
```

Comment:
Few more scenarios we can add:
- execute rule failure with revoked ephemeral key
- Testing a rule with spending limit 0 success case.

Reply:
1. Added a test case for executing rule with revoked key.
2. Ticket #49 is going to handle this case.